1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,140 @@
+
+- `set -e` - terminates on command error
+- `set -o pipefail` - returns an error if any command in the pipeline fails.
+
+> [!NOTE] A common configuration command for POSIX shells `set -euo pipefail` includes the `set -u`
+> configuration which returns an error if any variable has not been previously defined. This is
+> equivalent to the existing PowerShell `Set-StrictMode` and is not addressed in this RFC.
+
+The `$PSNativeCommandErrorAction` preference variable will implement a version of the
+`$ErrorActionPreference` variable for native commands.
+
+- The value will default to `Ignore` for compatibility with existing behavior.
+- For non-zero exit codes and except for the value `Ignore`, an `ErrorRecord` will be added to
+  `$Error` that wraps the exit code and the command executed that returned the exit code.
+- Initially, only the existing values of `$ErrorActionPreference` will be supported.
+- The set of values may be extended later to include `MatchErrorActionPreference`, which should
+  apply the `$ErrorActionPreference` setting to native commands.
+- A conversion will occur between `$PSNativeCommandErrorAction` and `$ErrorActionPreference`
+  values, where `MatchErrorActionPreference` is converted to the current value of
+  `$ErrorActionPreference`
+
+Valid values for `$PSNativeCommandErrorAction`
+
+| Value           | Definition
+----------------  | -------------------
+| Break           | Enter the debugger when an error occurs or when an exception is raised.
+| Continue        | (Default) - Displays the error message and continues executing.
+| Ignore          | Suppresses the error message and continues to execute the command.
+| Inquire         | Displays the error message and asks you whether you want to continue.
+| SilentlyContinue| No effect. The error message isn't displayed and execution continues without interruption.
+| Stop            | Displays the error message and stops executing. In addition to the error generated, the Stop value generates an ActionPreferenceStopException object to the error stream. stream
+| Suspend         | Automatically suspends a workflow job to allow for further investigation.
+
+The reported error record object will be the new type: `NativeCommandException` with the following details:
+
+| Property        | Definition
+----------------  | -------------------
+| ExitCode:      |  The exit code of the failed command.
+| ErrorID:       | `"Program {0} ended with non-zero exit code {1}"`, with the command name and the exit code, from resource string `ProgramFailedToComplete`.
+| ErrorCategory: | `ErrorCategory.NotSpecified`.
+| object:         | exit code
+| Source:        | The full path to the application
+| ProcessInfo    | details of failed command including path, exit code, and PID
+
+## Alternative Approaches and Considerations
+
+One way of checking for a single native command and handling its exit
```

Comment:
Include that on Windows, many native commands do not strictly follow the non-zero is error, so we may add a special list to handle those commands differently where known non-zero exit code is not treated as an error.

Reply:
@SteveL-MSFT, I suggest we not do that, because it adds complexity while providing little benefit: the "rogue" CLIs fall into two categories:

* True rogue CLIs that simply ignore exit codes and _always_ return `0`
  * There's nothing we can generally do to compensate for that, short of - brittlely - trying to analyze stderr output, which I don't think is worth doing.

* CLIs such as `robocopy.exe` that repurpose specific nonzero exit codes to communicate additional information  about _success_ conditions (while also using (other) nonzero exit codes to communicate failure).

 * These exit codes are CLI-specific, and the user needs to know about them and handle them; while we _could_ hard-code exceptions for the nonzero exit codes that signal _success_, the problems are:
    * Do we really want to code such exceptions for _all_ utilities that ship with Windows (assuming there are others - I don't know)? 
    * Users may expect the same magic to apply to _all_ utilities, whether in-box or not.
    * Maintenance burden: What if utilities later introduce _additional_ nonzero exit codes that aren't failures?

Instead, I think we should make it as easy as possible to ignore a nonzero exit code _ad hoc_, which can already be done, though it's not the most obvious solution:

```powershell
# Call robocopy and ignore any failure that a nonzero exit code may trigger.
robocopy ... || @()

# ... now analyze $LASTEXITCODE
```

Reply:
msiexec has two values that indicate success, but that a reboot is required (1641 and 3010), along with 0 also indicating success. That's the other one that I run into constantly (besides robocopy).

Discussed here: https://docs.microsoft.com/en-us/windows/win32/msi/error-codes

I'm sure there are more.

Reply:
Thanks, @essentialexch, that's good to know.

However, instead of getting into the hard-coded exception business, I suggest _documenting_ such high-profile exceptions as part of  a _conceptual help topic dedicated to external-program calls_: see https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5152